### DISCUSSION TITLE
Multi-class top-down ID model produces zero predictions during inference despite high training accuracy

### DISCUSSION THREAD

**Discussion by jonsen-lab** [talmolab/sleap#2709] [Help!] [answered]
Hello,
I am using SLEAP 1.6.3 with the sleap-nn 0.2.0 backend on Ubuntu Linux with a CUDA-enabled GPU. I am trying to track individual fish (3-4 per frame -> they can move in and out of frame) in top-down footage of murky water and general low visibility using the multi-animal top-down-id pipeline (centroid + Top Down ID models).
I used a Sigma of 10 for the centroid model and was playing around with sigma 5-7 in the ID-model. Epochs are set to 400 with a batch size of 4. Currently I have about 200 labelled frames and except for the individual tracking the node placement seems good so far.
Here is an example Frame of my video:
<img width="1828" height="854" alt="Screenshot from 2026-05-20 19-51-11" src="https://github.com/user-attachments/assets/c96508ed-d90c-4154-9fa5-d9048960687e" />
<img width="1349" height="758" alt="Screenshot from 2026-05-20 19-51-26" src="https://github.com/user-attachments/assets/e7b6443e-3e66-4796-a99f-cdc2c08cb951" />

Here are my Run Configurations
For the training:
<img width="1010" height="779" alt="Screenshot from 2026-05-20 20-16-26" src="https://github.com/user-attachments/assets/3e20f709-67e8-4915-9918-2c78dfe2a192" />
<img width="1025" height="778" alt="Screenshot from 2026-05-20 20-16-38" src="https://github.com/user-attachments/assets/18a9da36-7852-41c2-92be-5674ac9b57d3" />
<img width="1016" height="770" alt="Screenshot from 2026-05-20 20-16-50" src="https://github.com/user-attachments/assets/e7e18fd8-d51f-4c36-871d-c62e7a0a5cfc" />

For the Inference:
<img width="1042" height="744" alt="Screenshot from 2026-05-20 20-17-35" src="https://github.com/user-attachments/assets/dabcdadb-72b0-4af9-91dc-b142c5541b9f" />

The problem: Inference and training runs consistently produce zero predictions, both through the GUI (Predict menu) and via the terminal command sleap track, despite the models showing excellent training metrics.
Training metrics from the ID model (204 labelled frames):

Validation mAP: 98%
Validation PCK@5: 99%
Validation mOKS: ~100%
Node placement is clearly working well in the VIZ files

What I have tried:
Running inference on only labelled frames 
Running inference through the GUI with tracker set to simple
Testing the older centroid + centered_instance pipeline — also produces zero predictions
Running inference on the raw video file instead of the .slp labels file
Lowering --peak_threshold to 0.1 and 0.2 via terminal (I suspect the command didn't work or different problem despit not showing an error)

All runs through the SLEAP GUI completed without errors and finished in the expected time (4-8 h), but the output .slp file contains zero predictions every time. And after the run it says 0 instances on 204 frames. 
However, the tryouts via the termial (in order to try out a lower detection threshold) always seemed sketchy to me because they finished much quicker. (not the primary problem though but would be nice to know what I did wrong) 

Terminal command used:
sleap track --data_path /path/to/video.slp --only_labeled_frames --model_paths /path/to/centroid_model --model_paths /path/to/id_model --peak_threshold 0.1 /path/to/output.slp

Typical log output:
2026-05-20 | sleap-nn 0.2.0 | Python 3.13.5 | PyTorch 2.11.0+cu129 | CUDA 12.9 | 1 GPU(s)
2026-05-20 | Using device: cuda
Predicting... 100% 204/204 Elapsed: 0:01:19 2.0 FPS
2026-05-20 | Saved file at: ...
No errors are thrown. The output file simply contains no predictions. 

Questions:
1. Is there a known issue with the multi-animal top-down-id model producing zero predictions in SLEAP 1.6.3 / sleap-nn 0.2.0?
2. Is --peak_threshold actually supported for the multi-animal top-down-id pipeline, or is it silently ignored?
3. Is there a way to see what confidence scores the centroid model is actually producing, to determine whether detections are being made but filtered out?
4. Is there a way to work with negative frames on this or a different version of SLEAP while using this kind of model or are negative frames just not supported in these models yet?
5. Is the low confidence in detections really the issue why I don't see predictions in the predictions-file (only in VIZ files)?

Thanks alot!
Cheers Johannes

**Comment by tom21100227** [comment, marked as answer]
Hi @jonsen-lab , 

The number one issue I am seeing is **the receptive field of your centroid model**. I would change `max_stride` on your centroid model to where the blue square (receptive field) captures >= the size of those fish and retrain. 

https://nn.sleap.ai/dev/configuration/config-picker/ is a great tool to adjust all other parameters of your model. 

As for your questions: 

> Is there a known issue with the multi-animal top-down-id model producing zero predictions in SLEAP 1.6.3 / sleap-nn 0.2.0?

No, but in general id models are a bit more brittle than regular top-down models, so I would recommend retrain as a regular top down model if the parameter adjustment did not work. 

> Is --peak_threshold actually supported for the multi-animal top-down-id pipeline, or is it silently ignored?

Yes is it supported. Lowering threshold should produce more detected instance on average, but the issue here is that the receptive field is too small and your centroid model can't detect the fish. 

> Is there a way to see what confidence scores the centroid model is actually producing, to determine whether detections are being made but filtered out?

No through the GUI as of right now, but it is possible inside python. 
CLI:

```py
sleap-nn track --model_paths /path/to/centroid_model --data_path video.mp4
```

Python:
```py
from sleap_nn.predict import run_inference
run_inference(model_paths=["/path/to/centroid_model"], data_path="video.mp4")
```
or directly:
```py
from sleap_nn.inference.predictors import TopDownPredictor
predictor = TopDownPredictor.from_trained_models(
    centroid_ckpt_path="/path/to/centroid_model",
    confmap_ckpt_path=None,   # <-- this is what makes it centroid-only
)
predictor.make_pipeline("video.mp4")
output = predictor.predict(make_labels=True)
```

There is no built-in visualizer in sleap. Inference writes a .slp file containing sio.PredictedInstance objects where only the centroid coordinate is populated (other skeleton nodes are NaN). To visually see them you'd need to load them with sleap-io and plot yourself for now. 

> Is there a way to work with negative frames on this or a different version of SLEAP while using this kind of model or are negative frames just not supported in these models yet?

Negative frame just got supported in the [lastest sleap-nn](https://github.com/talmolab/sleap-nn/pull/484), you need to put `use_negative_frames=True` when starting the training. Please note this feature is fairly new so if there's any friction in getting it working we'd love any feedbacks! 

How you would edit your yaml files (training config): 
```yaml
...
data_config:
    use_negative_frames: true      # include user-confirmed negative frames
    negative_loss_weight: 1.0      # (optional) >1 amplifies negative gradient, <1 dampens
 ```

> Is the low confidence in detections really the issue why I don't see predictions in the predictions-file (only in VIZ files)?

Most definitely yes. 

Hope this helps! 

Best,
Tom 
Talmo Lab

**Reply by jonsen-lab** [reply]
Thanks alot for your quick answer. I will try the solutions you recommended.

You're the Best!
